sklearn_numba_dpex/kmeans/drivers.py

```diff
@@ -383,21 +401,69 @@ def _lloyd(
                 cluster_sizes_private_copies,
             )
 
+            if verbose:
+                # ???: verbosity comes at the cost of performance since it triggers
+                # computing exact inertia at each iteration. Shouldn't this be
+                # documented ?
+                # sklearn even goes further and re-computes inertia on updated clusters
+                # rather than old clusters but surely it probably isn't worth the
+                # computational cost of an additional pass on the data ?
+                inertia, *_ = dpctl.tensor.asnumpy(
+                    reduce_inertia_kernel(per_sample_inertia)
+                )
+                print(f"Iteration {n_iteration}, inertia {inertia:5.3e}")
+
             reduce_centroid_data_kernel(
                 cluster_sizes_private_copies,
                 new_centroids_t_private_copies,
                 cluster_sizes,
                 new_centroids_t,
                 empty_clusters_list,
-                nb_empty_clusters,
+                n_empty_clusters,
             )
 
-            nb_empty_clusters_ = dpctl.tensor.asnumpy(nb_empty_clusters)[0]
-            if nb_empty_clusters_ > 0:
-                # TODO: instead of this warning, set the centroid for empty clusters
-                # to new points carefully chosen in the dataset. (mimic scikit-learn
-                # behavior where the points with the highest inertia are chosen)
-                warnings.warn("Found an empty cluster", RuntimeWarning)
+            n_empty_clusters_ = int(n_empty_clusters[0])
+            if n_empty_clusters_ > 0:
+                # NB: empty cluster very rarely occurs, and it's more efficient to
+                # compute inertia and labels only after occurrences have been detected
+                # at the cost of an additional pass on data, rather than computing
+                # inertia by default during the first pass on data in case there's an
+                # empty cluster.
+                if verbose_and_use_uniform_weights:
+                    # if verbose and if sample_weight is uniform, distances to closest
+                    # centroids already have been computed in the main kernel we only
+                    # need to compute assignments
+                    fixed_window_assignment_kernel(
+                        X_t,
+                        centroids_t,
+                        centroids_half_l2_norm,
+                        assignments_idx,
+                    )
+                else:
+                    # else, compute distances + assignments
```

Comment:
```suggestion
                    # else, compute distances + assignments
                    # Note that we intentionally we pass unit weights instead of
                    # sample_weight so that per_sample_inertia will be updated
                    # to the (unweighted) squared distance to the
                    # nearest centroid.
```